backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inout/impl/HUInOutBL.java

```diff
@@ -373,7 +387,38 @@ public void copyAssignmentsToReversal(@NonNull final org.compiere.model.I_M_InOu
 				.stream()
 				.collect(ImmutableSetMultimap.toImmutableSetMultimap(
 						entry -> entry.getKey().getIdAssumingTableName(I_M_InOutLine.Table_Name, InOutLineId::ofRepoId),
-						entry -> entry.getValue()));
+						Map.Entry::getValue));
 	}
 
-}
+	@Override
+	public boolean isValidHuForReturn(final InOutId inOutId, final HuId huId)
+	{
+		final Optional<AttributeId> serialNoAttributeIdOptional = serialNoBL.getSerialNoAttributeId();
+		if (!serialNoAttributeIdOptional.isPresent())
+		{
+			return false;
+		}
+		final AttributeId serialNoAttributeId = serialNoAttributeIdOptional.get();
+
+		final I_M_HU hu = handlingUnitsDAO.getById(huId);
+		final I_M_HU_Attribute serialNoAttr = huAttributesDAO.retrieveAttribute(hu, serialNoAttributeId);
+		if (serialNoAttr == null)
+		{
+			//no S/N defined. Should not be a valid scenario
+			return false;
+		}
+		return !queryBL.createQueryBuilder(I_M_HU_Assignment.class).addOnlyActiveRecordsFilter()
+				.addEqualsFilter(I_M_HU_Assignment.COLUMNNAME_AD_Table_ID, inOutDAO.getInOutTableId())
```

Comment:
avoid using QueryBL outside repositories.

basically here, u shall do something like
* de.metas.handlingunits.inout.IHUInOutBL#getHUIdsByInOutLineIds - to retrieve the assigned HUIds. Consider introducing getHUIdsByInOutId
* use de.metas.handlingunits.impl.HandlingUnitsBL#getTopLevelHUs(java.util.Collection<de.metas.handlingunits.HuId>) - to get the top level HUIds
* having the top level HU Ids assigned to our InOut we can
* de.metas.handlingunits.impl.HandlingUnitsBL#createHUQueryBuilder
    * addOnlyHUIds(out top level HU Ids)
    * addOnlyWithAttribute(AttributeConstants.ATTR_SerialNo, "THE_SERIAL_NO")
* check if there is any match.

Reply:
refactored as suggested.